src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesGraphProvider.cs

```diff
@@ -0,0 +1,653 @@
+                                             SVsServiceProvider serviceProvider)
+            : base(new JoinableTaskContextNode(ThreadHelper.JoinableTaskContext))
+        {
+            ProjectContextProvider = projectContextProvider;
+            ServiceProvider = serviceProvider;
+        }
+
+        /// <summary>
+        /// Remembers expanded graph nodes.
+        /// TODO: Check how to clean them when project is unloaded. (What graph is doing in general when project is killed?)
+        /// </summary>
+        private WeakCollection<IGraphContext> ExpandedGraphContexts { get; set; }
+        private HashSet<string> RegisteredSubTreeProviders { get; set; }
+
+        private object _RegisteredSubTreeProvidersLock = new object();
+
+        private object _ExpandedGraphContextsLock = new object();
+
+        /// <summary>
+        /// All icons that are used tree graph, register their monikers once and refer to them by string id
+        /// </summary>
+        private HashSet<ImageMoniker> Images { get; set; }
+
+        protected override Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            ExpandedGraphContexts = new WeakCollection<IGraphContext>();
+            RegisteredSubTreeProviders = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            Images = new HashSet<ImageMoniker>();
+
+            ProjectContextProvider.ProjectContextChanged += ProjectContextProvider_ProjectContextChanged;
+
+            return TplExtensions.CompletedTask;
+        }
+
+        protected override Task DisposeCoreAsync(bool initialized)
+        {
+            ProjectContextProvider.ProjectContextChanged -= ProjectContextProvider_ProjectContextChanged;
+
+            return TplExtensions.CompletedTask;
+        }
+
+        private async Task EnsureInitializedAsync()
+        {
+            lock (_initializationLock)
+            {
+                if (_IsInitialized)
+                {
+                    return;
+                }
+
+                _IsInitialized = true;
+            }
+
+            await InitializeAsync().ConfigureAwait(false);
+        }
+
+        #region IGraphProvider
+
+        /// <summary>
+        /// IGraphProvider.BeginGetGraphData
+        /// Entry point for progression. Gets called everytime when progression
+        ///  - Needs to know if a node has children
+        ///  - Wants to get children for a node
+        ///  - During solution explorer search
+        /// </summary>
+        void IGraphProvider.BeginGetGraphData(IGraphContext context)
+        {
+            ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await BeginGetGraphDataAsync(context).ConfigureAwait(false);
+            });
+        }
+
+        /// <summary>
+        /// IGraphProvider.GetCommands
+        /// </summary>
+        IEnumerable<GraphCommand> IGraphProvider.GetCommands(IEnumerable<GraphNode> nodes)
+        {
+            yield return new GraphCommand(
```

Comment:
How often does this get called? This a 3 or 4 allocations every time it's called.